smile/socket_interface.py

```diff
@@ -0,0 +1,156 @@
+    ----------
+    unique_id : string
+        Unique identifier for this socket. Used only to differenciate this
+        socket from other sockets initialized with this funciton.
+    server : string
+        Host name. For a local TCP process, use the string "localhost"
+    port : int
+        A port number for the server.
+
+    Returns a Socket that is to be used in conjunction with the *SocketPush*
+    state. Will return None if the connection could not be established.
+    """
+
+
+    global _sockets
+
+    unique_identifier = "_".join([uniq_ID, server, str(port)])
+
+    if unique_identifier in _sockets.keys():
+        return _sockets[unique_identifier]
+
+    else:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        try:
+            rtn = sock.connect((server, port))
+            _sockets[unique_identifier] = sock
+            return _sockets[unique_identifier]
+
+        except:
+            sys.stdout.write("[ERROR  ] [SOCKET      ] Unable to establish a connection for the socket at server %s and the port %i\n"%(server, port))
+            return None
+
+
+class SocketPush(CallbackState):
+    """Pushes a message to the provided socket.
+
+    A state that uses a preinitialized Socket to send a message to a specific
+    server and port.
+
+    socket : socket
+        Preinitialized socket class using *init_socket_outlet*
+    msg : string
+        A string message that is to be passed to the specific socket.
+
+    Logged Attributes
+    -----------------
+    All parameters above and below are available to be accessed and
+    manipulated within the experiment code, and will be automatically
+    recorded in the state-specific log. Refer to State class
+    docstring for additional logged parameters.
+
+    send_time : Float
+        The time in which the message has finished being sent to the server.
+        This value is in seconds and based on experiment time. Experiment time
+        is the number of seconds since the start of the experiment.
+    msg : string
+        The message sent to the server.
+    server_name : string
+        The server the message is being sent to.
+    port : int
+        The port the message is being sent to in the server.
+    rtn : int
+        Number of bytes sent by this SocketPush. Check this value to make sure
+        your entire message was sent.
+
+    """
+
+
+    def __init__(self, socket, msg,  **kwargs):
+        super(SocketPush, self).__init__(parent=kwargs.pop("parent", None),
+                                         repeat_interval=kwargs.pop("repeat_interval", None),
+                                         duration=kwargs.pop("duration", 0.0),
+                                         save_log=kwargs.pop("save_log", True),
+                                         name=kwargs.pop("name", None),
+                                         blocking=kwargs.pop("blocking", True))
+        self._init_socket = socket
+        if socket is not None:
```

Comment:
Shouldn't this part filling out the socket stuff be in an enter, since you are technically allowing it to fill it from a ref?